core/src/main/clojure/xtdb/operator/window.clj

```diff
@@ -0,0 +1,237 @@
+                      :from-column ::lp/column)
+        :binary (s/cat :f simple-symbol?
+                       :left-column ::lp/column
+                       :right-column ::lp/column)))
+
+(s/def ::window-agg ::window-agg-expr)
+(s/def ::window-projection (s/map-of ::lp/column (s/keys :req-un [::window-name ::window-agg])))
+(s/def ::projections (s/coll-of ::window-projection :min-count 1))
+
+(defmethod lp/ra-expr :window [_]
+  (s/cat :op #{:window}
+         :specs (s/keys :req-un [::windows ::projections])
+         :relation ::lp/ra-expression))
+
+(comment
+  (s/valid? ::lp/logical-plan
+            '[:window {:windows {window1 {:partition-cols [a]
+                                          :order-specs [[b]]
+                                          :frame {}}}
+                       :projections [{col-name {:window-name window1
+                                                :window-agg (row-number)}}]}
+              [:table [{}]]]))
+
+
+#_{:clj-kondo/ignore [:unused-binding :clojure-lsp/unused-public-var]}
+(definterface IWindowFnSpec
+  (^void aggregate [^org.apache.arrow.vector.IntVector groupMapping
+                    ^ints sortMapping
+                    ^xtdb.vector.RelationReader in-rel])
+  (^xtdb.vector.IVectorReader finish []))
+
+#_{:clj-kondo/ignore [:unused-binding :clojure-lsp/unused-public-var]}
+(definterface IWindowFnSpecFactory
+  (^clojure.lang.Symbol getToColumnName [])
+  (getToColumnField [])
+  (^xtdb.operator.window.IWindowFnSpec build [^org.apache.arrow.memory.BufferAllocator allocator]))
+
+#_{:clj-kondo/ignore [:unused-binding]}
+(defmulti ^xtdb.operator.window.IWindowFnSpecFactory ->window-fn-factory
+  (fn [{:keys [f from-name from-type to-name zero-row?]}]
+    (expr/normalise-fn-name f)))
+
+(defmethod ->window-fn-factory :row_number [{:keys [to-name]}]
+  (reify IWindowFnSpecFactory
+    (getToColumnName [_] to-name)
+    (getToColumnField [_] (types/col-type->field :i64))
+
+    (build [_ al]
+      (let [^BigIntVector out-vec (-> (types/col-type->field to-name :i64)
+                                      (.createVector al))
+            ;; TODO this can become large if groups are small
+            ^HashMap group-to-cnt (HashMap.)]
+        (reify
+          IWindowFnSpec
+          (aggregate [_ group-mapping sortMapping _in-rel]
+            (let [offset (.getValueCount out-vec)
+                  row-count (.getValueCount group-mapping)]
+              (.setValueCount out-vec (+ offset row-count))
+              (dotimes [idx row-count]
+                (.set out-vec (+ offset idx)
+                      ^long (.compute group-to-cnt
+                                      (.get group-mapping (aget sortMapping idx))
+                                      (fn [_group-idx cnt] (or (some-> cnt inc) 0)))))))
+
+          (finish [_]
+            (vr/vec->reader out-vec))
+
+          Closeable
+          (close [_] (.close out-vec)))))))
+
+(doseq [op [:min :max]]
```

Comment:
worth chucking this in a `comment` block if it's currently just at 'hack' stage so the next person understands the situation :slightly_smiling_face: 

(I haven't reviewed the impl)